tools/shared/src/main/scala/org/scalajs/core/tools/optimizer/IRFileCache.scala

```diff
@@ -0,0 +1,155 @@
+/*                     __                                               *\
+**     ________ ___   / /  ___      __ ____  Scala.js tools             **
+**    / __/ __// _ | / /  / _ | __ / // __/  (c) 2013-2014, LAMP/EPFL   **
+**  __\ \/ /__/ __ |/ /__/ __ |/_// /_\ \    http://scala-js.org/       **
+** /____/\___/_/ |_/____/_/ | |__/ /____/                               **
+**                          |/____/                                     **
+\*                                                                      */
+
+
+package org.scalajs.core.tools.optimizer
+
+import scala.collection.mutable
+
+import org.scalajs.core.tools.logging._
+import org.scalajs.core.tools.io._
+
+import org.scalajs.core.ir
+import ir.Infos
+
+final class IRFileCache {
+
+  private[this] val irCache = mutable.Map.empty[String, PersistentIRFile]
+
+  private[this] var readyToUse: Boolean = false
+
+  private[this] var statsReused: Int = 0
+  private[this] var statsInvalidated: Int = 0
+  private[this] var statsTreesRead: Int = 0
+
+  /** Updates this cache with new IR files.
+   *
+   *  Successive calls of this method must be interlieved with calls to
+   *  [[cleanAfterUse]].
+   *  @returns true if a file or the composition of files changed
+   */
+  def update(irFiles: Traversable[VirtualScalaJSIRFile]): Boolean = {
+    requireReadyState(false, "Call cleanAfterUse before calling update again")
+    readyToUse = true
+
+    // Are we sure we'll change something based on file count?
+    val baseUpdated = irFiles.size != irCache.size
+
+    // Update every file
+    irFiles.foldLeft(baseUpdated) { (updated, irFile) =>
+      val file = irCache.getOrElseUpdate(irFile.path,
+          new PersistentIRFile(irFile.path))
+      val thisUpdated = file.updateFile(irFile) 
+      updated || thisUpdated
+    }
+  }
+
+  /** Cleans this cache after usage.
+   * 
+   *  You must call this method before calling [[update]] again.
+   *  @returns Caching statistics about this use
+   */
+  def cleanAfterUse(): IRFileCache.Stats = {
+    requireReadyState(true, "Cache already clean or uninitialized. Call update")
+    readyToUse = false
+
+    val resultStats =
+      new IRFileCache.Stats(statsReused, statsInvalidated, statsTreesRead)
+
+    // "Garbage-collect" persisted versions of files that have disappeared
+    irCache.retain((_, f) => f.cleanAfterUse())
```

Comment:
Yes. We could do that. I first thought it restricts us from changing the caching strategy, but it doesn't seem to actually do so.
The only downside is that you cannot clean up the cache without updating it. So we might use more memory during the late phase of the pipeline (since we retain everything until the next run). Basically this means this patch potentially *increases* memory usage. Your call.